copy_from.go

```diff
@@ -166,20 +175,32 @@ func (ct *copyFrom) run(ctx context.Context) (int64, error) {
 	return commandTag.RowsAffected(), err
 }
 
-func (ct *copyFrom) buildCopyBuf(buf []byte, sd *pgconn.StatementDescription) (bool, []byte, error) {
+func (ct *copyFrom) buildCopyBuf(buf []byte, oids []uint32) (bool, []byte, error) {
+
+	if oids != nil && len(ct.columnNames) != len(oids) {
+		return false, nil, fmt.Errorf("expected %d OIDs, got %d OIDs", len(ct.columnNames), len(oids))
+	}
 
 	for ct.rowSrc.Next() {
 		values, err := ct.rowSrc.Values()
 		if err != nil {
 			return false, nil, err
 		}
+
 		if len(values) != len(ct.columnNames) {
 			return false, nil, fmt.Errorf("expected %d values, got %d values", len(ct.columnNames), len(values))
 		}
 
+		if oids == nil {
+			oids, err = ct.oidsFromValues(values)
+			if err != nil {
+				return false, nil, fmt.Errorf("couldn't get OIDs from values: %w", err)
+			}
+		}
```

Comment:
Here I check if OIDs were pass as arguments, if they weren't, which is the case for `case QueryExecModeExec, QueryExecModeSimpleProtocol` then It defaults to matching the types of the values against the `pgtype.Map`. I need to do this here because I can't peek at the values from `rowSrc` without moving the iterator. 

I'll add some comments for this path to make it more explicit.

Reply:
I added the extra comments.

Reply:
Ah. I don't think this will work for types that are not exactly 1 to 1 for Go to PostgreSQL. e.g. a Go `time.Time` and a PostgreSQL `date` or Go `int64` and PostgreSQL `smallint`.

It previously would run a prepare even in `QueryExecModeExec` and `QueryExecModeExec` to avoid these problems. I don't think that can safely be removed.

Reply:
You mean because of: https://github.com/jackc/pgx/blob/e58381ac9409172a674f76c1801c3c606ac41b42/extended_query_builder.go#L148-L160

I just need a way to specify the OIDs for the copy operation. I was thinking about:

> If it really needs to be controlled per call, perhaps a new optional interface could be created that rowSrc could implement. This interface could have a method to get the query exec mode. A wrapper struct could be created that wrapped the original CopyFromSource and added the new method.

It would be ideal if I can extend the `CopyFromSource` with a method that returns the OIDs of the values, but that would mean changing the interface. We could try the wrapper that you mentioned, but then at some point in the code I'll need to cast the `CopyFromSource` to check if it's a wrapper struct, and then call the function, which might not be the best solution. Something like changing 

https://github.com/jackc/pgx/pull/1412/files#diff-89f7d5816796e2870b38b63109c1f9a3f9a382420f245dffcada921d3878760cR109

For:

```go
var oids []uint32
w, ok := (ct.rowSrc).(withOIDsWrapper)
if ok {
         oids = w.OIDs()
} else {
	switch ct.mode {
	case QueryExecModeExec, QueryExecModeSimpleProtocol:
                 ct.mode = QueryExecModeDescribeExec
                 fallthrough
	case QueryExecModeCacheStatement, QueryExecModeCacheDescribe, QueryExecModeDescribeExec:
		var err error
		oids, err = ct.oidsFromStatementDescription(ctx, quotedColumnNames, quotedTableName)
		if err != nil {
			return 0, err
		}
	default:
		return 0, fmt.Errorf("unknown QueryExecMode: %v", ct.mode)
}
```

If this is something that you don't want to add, I can modify the PR to always use the defaultExecMode and to handle the cases of `QueryExecModeExec` and `QueryExecModeSimpleProtocol`, default to use `QueryExecModeDescribeExec` because it's the same behaviour as before. For our use case, we could just construct our own CopyFrom using `pgconn.CopyFrom` directly.

Let me know what you think, or if you have other ideas. Thanks :)

Reply:
I think this is the correct approach. i.e. `QueryExecModeDescribeExec`, `QueryExecModeExec`, `QueryExecModeSimpleProtocol` would all call `Prepare()` like they do now and only `QueryExecModeCacheStatement` and `QueryExecModeCacheDescribe` would use the new path.

That would finish this unit of work. Then in another optional PR we could work out that extra interface for `CopyFromSource` to manually control this behavior.

Reply:
I'll try to get this donde by the end of the week, otherwise it'll have to wait until next year. Thanks for the input :)